libnopegl/src/text_external.c

```diff
@@ -0,0 +1,749 @@
+            if (ret < 0)
+                return ret;
+
+            /* Save the rasterized glyph in the index */
+            struct glyph *glyph = create_glyph();
+            if (!glyph)
+                return NGL_ERROR_MEMORY;
+            glyph->atlas_id  = atlas_id;
+            glyph->width     = NGLI_I32_TO_I26D6(bitmap.width);
+            glyph->height    = NGLI_I32_TO_I26D6(bitmap.height);
+            glyph->bearing_x = NGLI_I32_TO_I26D6(slot->bitmap_left);
+            glyph->bearing_y = NGLI_I32_TO_I26D6(slot->bitmap_top - bitmap.height);
+            ret = ngli_hmap_set(glyph_index, glyph_uid, glyph);
+            if (ret < 0) {
+                free_glyph(NULL, glyph);
+                return ret;
+            }
+        }
+    }
+
+    return 0;
+}
+
+static void reset_runs(struct darray *runs_array)
+{
+    struct text_run *runs = ngli_darray_data(runs_array);
+    for (size_t i = 0; i < ngli_darray_count(runs_array); i++)
+        hb_buffer_destroy(runs[i].buffer);
+    ngli_darray_reset(runs_array);
+}
+
+/* Make sure we can hold the whole Unicode codepoints */
+NGLI_STATIC_ASSERT(fribidi_chars_are_32bit, sizeof(FriBidiChar) == 4);
+
+static int char_is_linebreak(FriBidiChar ch)
+{
+    /* Source: https://en.wikipedia.org/wiki/Whitespace_character */
+    switch (ch) {
+    case 0x000A: /* line feed */
+    case 0x000B: /* line tabulation */
+    case 0x000C: /* form feed */
+    case 0x000D: /* carriage return */
+    case 0x0085: /* next line */
+    case 0x2028: /* line separator */
+    case 0x2029: /* paragraph separator */
+        return 1;
+    }
+    return 0;
+}
+
+static int char_is_whitespace(FriBidiChar ch)
+{
+    return fribidi_get_bidi_type(ch) == FRIBIDI_TYPE_WS;
+}
+
+static size_t find_line_end(const FriBidiChar *str, size_t len, size_t start)
+{
+    for (size_t i = start; i < len; i++)
+        if (char_is_linebreak(str[i]))
+            return i;
+    return len;
+}
+
+static int append_run(struct text *text, struct darray *runs_array,
+                      const FriBidiChar *str, size_t len, size_t face_id,
+                      enum run_type type, size_t start, size_t end)
+{
+    hb_buffer_t *buffer = hb_buffer_create();
+    if (!hb_buffer_allocation_successful(buffer))
+        return NGL_ERROR_MEMORY;
+
+    hb_buffer_add_codepoints(buffer, str, (int)len, (unsigned)start, (int)(end - start));
+
+    if (text->config.writing_mode == NGLI_TEXT_WRITING_MODE_VERTICAL_LR ||
+        text->config.writing_mode == NGLI_TEXT_WRITING_MODE_VERTICAL_RL) {
+        hb_buffer_set_direction(buffer, HB_DIRECTION_TTB);
+    } else if (text->config.writing_mode == NGLI_TEXT_WRITING_MODE_HORIZONTAL_TB) {
+        /* FriBidi changes the codepoints order from right-to-left where appropriate */
+        hb_buffer_set_direction(buffer, HB_DIRECTION_LTR);
+    } else {
+        ngli_assert(0);
+    }
+
+    // TODO: expose to the user?
+    //hb_buffer_set_script(buffer, script);
```

Comment:
Can you explain a bit more in the comment how that woul be useful ?
Also can you add a space between `//` and `hb_`

Reply:
Changed to:
```c
    // TODO: expose to the user? The precise knowledge of the language and/or
    // script may affect the choice of symbols, rules and conventions used
    // during shaping.
    // hb_buffer_set_script(buffer, script);
    // hb_buffer_set_language(buffer, hb_language_from_string("en", -1));
```

Reply:
Thanks